cproto.1
.\" $Id: cproto.1,v 4.9 2010/07/14 09:43:24 tom Exp $
.\"
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +5n
..
.de EE		\"End example
.fi
.in -5n
.if n .sp 1
.if t .sp .5
..
.TH CPROTO 1 "July 2010"
.
.SH NAME
cproto \- generate C function prototypes and convert function definitions
.
.SH SYNOPSIS
.B cproto
[ 
.I option \fP...\fI
] [
.I file \fP...\fI
]
.
.SH DESCRIPTION
.B Cproto
generates function prototypes for
functions defined in the specified C source files to the standard output.
The function definitions may be in the old style or ANSI C style.
Optionally,
.B cproto
also outputs declarations for variables defined in the files.
If no
.I file
argument is given,
.B cproto
reads its input from the standard input.
.LP
By giving a command line option,
.B cproto
will also convert function definitions in the specified files from the
old style to the ANSI C style.
The original source files along with files specified by
.EX
#include "file"
.EE
directives appearing in the source code will be overwritten
with the converted code.
If no file names are given on the command line, then the program reads
the source code from the standard input and outputs the converted source
to the standard output.
.LP
If any comments appear in the parameter declarations for a function definition,
such as in the example,
.EX
main (argc, argv)
int argc;       /* number of arguments */
char *argv[];   /* arguments */
{
}
.EE
then the converted function definition will have the form
.EX
int
main (
    int argc,       /* number of arguments */
    char *argv[]   /* arguments */
)
{
}
.EE
Otherwise, the converted function definition will look like
.EX
int
main (int argc, char *argv[])
{
}
.EE
.LP
.B Cproto
can optionally convert function definitions from the ANSI style to the old
style.
In this mode, the program also converts function declarators and prototypes
that appear outside function bodies.
This is not a complete ANSI C to old C conversion.
The program does not change anything within function bodies.
.LP
.B Cproto
can optionally generate source in lint\-library format.
This is useful in environments where the \fBlint\fR utility is
used to supplement prototype checking of your program.
.
.SH OPTIONS
.TP
.B \-e
Output the keyword
.B extern
in front of every generated prototype or declaration that has global scope.
.TP
.BI \-f " n"
Set the style of generated function prototypes where
.I n
is a number from 0 to 3.
For example, consider the function definition
.EX
main (argc, argv)
int argc;
char *argv[];
{
}
.EE
If the value is 0, then no prototypes are generated.
When set to 1, the output is:
.EX
int main(/*int argc, char *argv[]*/);
.EE
For a value of 2, the output has the form:
.EX
int main(int /*argc*/, char */*argv*/[]);
.EE
The default value is 3.
It produces the full function prototype:
.EX
int main(int argc, char *argv[]);
.EE
.TP
.B \-l
Generate text for a lint\-library (overrides the "\fB\-f\fR" option).
The output includes the comment
.EX
/* LINTLIBRARY */
.EE
Special comments LINT_EXTERN and LINT_PREPRO (a la "VARARGS")
respectively turn on the "\-x"
option and copy comment\-text to the output (for preprocessing in \fBlint\fR).
Use the comment
.EX
/* LINT_EXTERN2 */
.EE
to include externs defined in the first level of include\-files.
Use the comment
.EX
/* LINT_SHADOWED */
.EE
to cause 
.B cproto
to put "#undef" directives before each lint library declaration
(i.e., to avoid conflicts with macros that happen to have to have
the same name as the functions, thus causing syntax errors).
.P
Note that these special comments
are not supported under VAX/VMS, since there is no
equivalent for the "\-C" option of cpp with VAX\-C.
.TP
.B \-c
The parameter comments in the prototypes generated by
the \-f1 and \-f2 options are omitted by default.
Use this option to enable the output of these comments.
.TP
.B \-m
Put a macro around the parameter list of every generated prototype.
For example:
.EX
int main P_((int argc, char *argv[]));
.EE
.TP
.BI \-M " name"
Set the name of the macro used to surround prototype parameter lists
when option \-m is selected.
The default is "P_".
.TP
.B \-d
Omit the definition of the prototype macro used by the \-m option.
.TP
.BI \-o " file"
Specify the name of the output file (default: standard output).
.TP
.BI \-O " file"
Specify the name of the error file (default: standard error).
.TP
.B \-p
Disable promotion of formal parameters in old style function definitions.
By default, parameters of type
.B char
or
.B short
in old style function definitions are promoted to type
.B int
in the function prototype or converted ANSI C function definition.
Parameters of type
.B float
get promoted to 
.B double
as well.
.TP
.B \-q
Do not output any error messages when the program cannot read the
file specified in an
.I #include
directive.
.TP
.B \-s
By default,
.B cproto
only generates declarations for functions and variables having global scope.
This option will output
.B static
declarations as well.
.TP
.B \-S
Output only static declarations.
.TP
.B \-i
By default,
.B cproto
only generates declarations for functions and variables having global scope.
This option will output
.B inline
declarations as well.
.TP
.B \-T
Copy type definitions from each file.
(Definitions in included\-files are copied, unlike the "\-l" option).
.TP
.B \-v
Also output declarations for variables defined in the source.
.TP
.B \-x
This option causes procedures and variables which are declared "extern"
to be included in the output.
.TP
.BI \-X " level"
This option limits the include\-file level from which declarations are extracted
by examining the preprocessor output.
.TP
.B \-a
Convert function definitions from the old style to the ANSI C style.
.TP
.B \-t
Convert function definitions from the ANSI C style to the traditional style.
.TP
.B \-b
Rewrite function definition heads to include both old style and new style
declarations separated by a conditional compilation directive.
For example, the program can generate this function definition:
.EX
#ifdef ANSI_FUNC

int
main (int argc, char *argv[])
#else

int
main (argc, argv)
int argc;
char *argv[]
#endif
{
}
.EE
.TP
.BI \-B " directive"

CHANGES
-- $Id: CHANGES,v 4.41 2014/01/01 15:46:55 tom Exp $

2014/01/01 (4.7l)
- add "docs" rule to makefile.
- correct configure check for yyname vs yytname, broken by 4.7i changes
- update config.guess

2013/10/25 (4.7k)
- modify configure script, etc., to support cross-compiles, e.g., to MinGW.
- updated configure macros
  - support --datarootdir option
  - various portability fixes for CF_XOPEN_SOURCE
  - add checks for clang
  - improve macros for checking for standard C compiler 
  - modified macros to support third parameter of AC_DEFINE used by autoheader
- update config.guess, config.sub

2011/01/02 (4.7j)
- amend a check for function name to allow cases such as this:
  int (foo)();
  which is distinct from a function-pointer, e.g.,
  int (*foo)();
- modified configure script to deprecate ${name-value} in favor of standard
  ${name:-value}.
- update config.guess, config.sub

2010/07/14 (4.7i)
- escape dashes used in manpage, to work with groff (patch by Kenneth
  Pronovici) 
- add RPM and Debian build scripts, for testing.
- add configure checks for ctags and lint program.
- fix strict compiler warnings, e.g., using const.
- support DESTDIR variable in makefile.
- drop mkdirs.sh, use "mkdir -p"
- several improvements to configure macros
  - quoted ifelse() params
  - distinguish Darwin 9 from previous releases
  - ignore stderr in check for gcc version, to work with c89 wrapper.
- use consistent date and formatting in the CHANGES file, for processing
  contributor's information using a script.
- update config.guess, config.sub

2009/09/03 (4.7h)
- updated configure macros CF_ADD_CFLAGS CF_GCC_ATTRIBUTES CF_GCC_WARNINGS
  CF_XOPEN_SOURCE
- update config.guess, config.sub

2008/11/19 (4.7g)
- modify cpp command to redirect stderr to /dev/null if -q option is given,
  for consistency with the non-cpp mode (report by Nicolas Limare,
  Ubuntu #275248).
- add configure --disable-leaks option.
- use configure macro CF_XOPEN_SOURCE macro to make mkstemp() prototyped on
  Linux.
- remove isascii() usage.
- code cleanup, to remove K&R relics.
- update config.guess, config.sub

2008/01/01 (4.7f)
- add symbol for __gnuc_va_list
- add some data for c99 syntax to test-cases, e.g., long long.
- review/fix some additional places where need_temp() call is needed.
- fix a reference to unallocated storage when reading from standard input
  (Fedora #315061).
- minor updates to configure script macros
- update mkdirs.sh (for parallel makes)
- update config.guess, config.sub
- rename install.sh to install-sh

2005/12/08 (4.7e)
- eliminate some fixed limits on buffer sizes (prompted by FreeBSD port).
- eliminate fixed limit on include nesting.
- use configure check for mkstemp(), use that function in preference to
  mktemp() if a working version is found.
- move strstr.c into strkey.c to avoid zero-length object
- improve configure check for gcc version, from ncurses.
- update config.guess, config.sub

2005/08/21 (4.7d)
- modified configure script and makefile to work with cygwin
- fix configure check for yacc errors broken by 4.7c changes.

cproto.c
/* $Id: cproto.c,v 4.36 2014/01/01 15:30:44 tom Exp $
 *
 * C function prototype generator and function definition converter
 */
#define VERSION "4.7l.1.1"

#include <cproto.h>

/* getopt declarations */
#if HAVE_GETOPT_H
#include <getopt.h>
#else
extern int getopt(int argc, char *const *argv, const char *shortopts);
extern char *optarg;
extern int optind;
#endif

/* Name of the program (from argv[0]) */
char *progname;

/* Program options */

/* If nonzero, output variables declared "extern" in include-files */
unsigned extern_in = 0;

/* When TRUE, track the include-level (works with gcc, not some others) */
#if OPT_LINTLIBRARY
int do_tracking = FALSE;
#endif

/* When TRUE, suppress return-statements in function-bodies */
int exitlike_func = FALSE;

/* If TRUE, output "extern" before global declarations */
boolean extern_out = FALSE;

/* By default, generate global declarations only */
Scope scope_out = SCOPE_EXTERN;

/* If TRUE, output definitions for inline functions */
boolean inline_out = FALSE;

/* If TRUE, export typedef declarations (needed for lint-libs) */
#if OPT_LINTLIBRARY
boolean types_out = FALSE;
#endif

/* If TRUE, undef functions to avoid shadowing problems */
#if OPT_LINTLIBRARY
boolean lint_shadowed = FALSE;
#endif

/* If TRUE, search the standard system directories for header files */
boolean use_stdinc_dirs = TRUE;

/* If TRUE, generate variable declarations */
boolean variables_out = FALSE;

/* If TRUE, enable formal parameter promotion */
boolean promote_param = TRUE;

/* Style of function prototype to generate */
PrototypeStyle proto_style = PROTO_ANSI;

/* Function definition style converted to */
FuncDefStyle func_style = FUNC_UNKNOWN;

/* If TRUE, put guard macro around prototype parameters */
boolean proto_macro = FALSE;

/* Name of macro to guard prototypes */
const char *macro_name = "P_";

/* If TRUE, output prototype macro definition */
boolean define_macro = TRUE;

/* If TRUE, output comments in prototypes */
boolean proto_comments = FALSE;

/* If TRUE, output comments naming source files */
boolean file_comments = TRUE;

/* Conditional compilation directive output in front of function definitions */
const char *func_directive = "#ifdef ANSI_FUNC";

/* Output formats for function declarators */
FuncFormat fmt[] =
{
    /* miscellaneous function declarator */
    {"", " ", "", "", " ", ""},
    /* prototype */
    {"", " ", "", "", " ", ""},
    /* function definition */
    {"", "\n", " ", "", " ", ""},
    /* function definition with parameter comments */
    {"", "\n", " ", "\n    ", "\n    ", "\n"},
};

/* If TRUE, don't output message if unable to read an include file */
boolean quiet = FALSE;

/* Include file directories */
unsigned num_inc_dir = 0;
char **inc_dir = 0;

/* Run the C preprocessor */
#ifdef CPP
# if !HAVE_POPEN_PROTOTYPE
extern FILE *popen(const char *c, const char *m);
extern int pclose(FILE *p);
# endif
static size_t cpp_len;
static const char *cpp = CPP;
static char *cpp_opt;
static char *cpp_cmd;
#endif

/* Try to allocate some memory.
 * If unsuccessful, output an error message and exit.
 */
#if !HAVE_LIBDMALLOC
#ifdef NO_LEAKS
void *
xMalloc(size_t n, char *f GCC_UNUSED, int l GCC_UNUSED)
#else
void *
xmalloc(size_t n)
#endif
{
    void *p;
#if HAVE_LIBDBMALLOC
    p = debug_malloc(f, l, n);
#else
    p = malloc(n);
#endif

    if (p == NULL) {
	fprintf(stderr, "%s: out of memory (cannot allocate %lu bytes)\n",
		progname, (unsigned long) n);
	exit(EXIT_FAILURE);
    }
    return p;
}
#endif /* if !HAVE_LIBDMALLOC */

/* Try to reallocate some memory.
 * If unsuccessful, output an error message and exit.
 */
#if !HAVE_LIBDMALLOC
#ifdef NO_LEAKS
void *
xRealloc(void *p, size_t n, char *f GCC_UNUSED, int l GCC_UNUSED)
#else
void *
xrealloc(void *p, size_t n)
#endif
{
#if HAVE_LIBDBMALLOC
    p = debug_malloc(f, l, p, n);
#else
    p = realloc(p, n);
#endif

    if (p == NULL) {
	fprintf(stderr, "%s: out of memory (cannot allocate %lu bytes)\n",
		progname, (unsigned long) n);
	exit(EXIT_FAILURE);
    }
    return p;
}
#endif /* if !HAVE_LIBDMALLOC */

/* Copy the string into allocated memory.
 * If unsuccessful, output an error message and exit.
 */
#if !HAVE_LIBDMALLOC
#ifdef NO_LEAKS
char *
xStrdup(const char *src, char *f, int l)
#else
char *
xstrdup(const char *src)
#endif
{
#if defined(NO_LEAKS)
    return strcpy((char *) xMalloc(strlen(src) + 1, f, l), src);
#else
    return strcpy((char *) xmalloc(strlen(src) + 1), src);
#endif
}
#endif /* if !HAVE_LIBDMALLOC */

/* Output the current source file name and line number.
 */
void
put_error(void)
{
    fprintf(stderr, "%s:%u: ", cur_file_name(), cur_line_num());
}

/* Scan for options from a string.
 */
static void
parse_options(char *src, int maxargc, int *pargc, char **argv)
{
    char *g, *p, c;
    int argc;

    argc = 0;
    g = xstrdup(src);
    c = *g;
    while (c != '\0' && argc < maxargc) {
	while (c == ' ' || c == '\t')
	    c = *++g;
	if (c == '\0')
	    break;
	argv[argc++] = g;

	p = g;
	for (;;) {
	    if (c == ' ' || c == '\t' || c == '\0') {
		*p = '\0';
		break;
	    } else if (c == '"') {
		for (;;) {
		    c = *++g;
		    if (c == '"') {
			c = *++g;
			break;
		    } else if (c == '\0') {
			break;
		    } else {
			*p++ = c;
		    }
		}
	    } else {
		*p++ = c;
		c = *++g;
	    }
	}
	if (c != '\0')
	    c = *++g;
    }

    *pargc = argc;
}

/* Replace any character escape sequences in a string with the actual
 * characters.	Return a pointer to malloc'ed memory containing the result.
 * This function knows only a few escape sequences.
 */
static char *
escape_string(char *src)
{
    char *result, *get, *put;

    result = xstrdup(src);
    put = result;
    get = src;
    while (*get != '\0') {
	if (*get == '\\') {
	    switch (*(++get)) {
	    case 'n':
		*put++ = '\n';
		++get;
		break;
	    case 's':
		*put++ = ' ';
		++get;
		break;
	    case 't':
		*put++ = '\t';
		++get;
		break;
	    default:
		if (*get != '\0')
		    *put++ = *get++;
	    }
	} else {
	    *put++ = *get++;
	}
    }
    *put = *get;
    return result;
}

/* Returns true iff the character is a path leaf separator
 */
int
is_path_sep(int ch)
{
#if defined(MSDOS) || defined(OS2)
    return ch == '/' || ch == '\\';
#else
    return ch == '/';
#endif
}

/* Trim any path name separator from the end of the string.
 * Return a pointer to the string.
 */
char *
trim_path_sep(char *s)
{
    size_t n = strlen(s);

    if (n != 0) {
	if (is_path_sep(s[n - 1]))
	    s[n - 1] = '\0';
    }
    return s;
}

/* Output usage message and exit.
 */
static void
usage(void)
{
    fprintf(stderr, "usage: %s [ option ... ] [ file ... ]\n", progname);
    fputs("Options:\n", stderr);
    fputs("  -a, -t           Convert function definitions to ANSI or traditional style\n", stderr);
    fputs("  -b               Rewrite function definitions in both styles\n", stderr);
    fputs("  -c               Enable comments in prototype parameters\n", stderr);
    fputs("  -e               Output \"extern\" keyword before global declarations\n", stderr);
    fputs("  -f n             Set function prototype style (0 to 3)\n", stderr);
#if OPT_LINTLIBRARY
    fputs("  -l               Generate output in lint-library style\n", stderr);
#endif
    fputs("  -o file          Redirect output to file\n", stderr);
    fputs("  -O file          Redirect errors to file\n", stderr);
    fputs("  -p               Disable formal parameter promotion\n", stderr);
    fputs("  -q               Disable include file read failure messages\n", stderr);
    fputs("  -s               Output static declarations also\n", stderr);
    fputs("  -S               Output static declarations only\n", stderr);
    fputs("  -i               Output inline declarations also\n", stderr);
#if OPT_LINTLIBRARY
    fputs("  -T               Output type definitions\n", stderr);
#endif
    fputs("  -v               Output variable declarations\n", stderr);
    fputs("  -x               Output variables and functions declared \"extern\"\n", stderr);
#if OPT_LINTLIBRARY
    fputs("  -X level         Limit externs to given include-level\n", stderr);
#endif
    fputs("  -m               Put macro around prototype parameters\n", stderr);
    fputs("  -M name          Set name of prototype macro\n", stderr);
    fputs("  -d               Omit prototype macro definition\n", stderr);
    fputs("  -P template      Set prototype format template \" int f (a, b)\"\n", stderr);
    fputs("  -F template      Set function definition format template \" int f (a, b)\"\n", stderr);
    fputs("  -C template      Set format for function definition with parameter comments\n", stderr);
    fputs("  -D name[=value]  Define C preprocessor symbol\n", stderr);
    fputs("  -U name          Undefine C preprocessor symbol\n", stderr);
    fputs("  -I directory     Add #include search directory\n", stderr);
    fputs("  -n               Do not search the standard system directories for header files\n", stderr);
    fputs("  -E command       Run specified C preprocessor command\n", stderr);
    fputs("  -E 0             Do not run any C preprocessor\n", stderr);
    fputs("  -V               Print version information\n", stderr);
    exit(EXIT_FAILURE);
}

#define CHUNK(n) (((n) | 7) + 1)

static void
add_inc_dir(const char *src)
{
    size_t have = CHUNK(num_inc_dir);
    size_t need = CHUNK(num_inc_dir + 1);
    size_t used = (need * sizeof(char *));

    if (inc_dir == 0) {
	inc_dir = (char **) malloc(used);
    } else if (have != need) {
	inc_dir = (char **) realloc(inc_dir, used);
    }

    inc_dir[num_inc_dir++] = trim_path_sep(xstrdup(src));
}

#ifdef	vms
static char *cpp_defines;
static char *cpp_include;
static char *cpp_undefns;

static void
add2list(char *dst, char *src)
{
    if (*dst)
	strcat(dst, ",");
    strcat(dst, src);
}

static void
add_option(char *keyword, char *src)
{
    if (*src)
	sprintf(cpp_opt + strlen(cpp_opt), " /%s=(%s)", keyword, src);
}
#endif /* vms */

#ifdef QUOTE_POPEN_ARGS

/* Calculate length of string including shell quoting characters that
 * must be inserted to preserve the string when it is passed to /bin/sh.
 * On UNIX systems, popen() runs /bin/sh.
 */
#define QUOTECHARS "\"\'\t\n "

static int
quote_length(char *s)
{
    int len = strlen(s);

    if (strpbrk(s, QUOTECHARS)) {
	len += 2;
	while (*s)
	    if (*s++ == '\'')
		len += 4;	/* replace ' with '"'"' (ick!) */
    }
    return len;
}

/* Insert quoting characters to preserve the string when it is passed to
 * /bin/sh.
 */
static char *
quote_string(char *s)
{
    if (strpbrk(s, QUOTECHARS)) {
	char *src = s;
	char *dup, *dup_orig;

	dup = dup_orig = xstrdup(s);

	while (isspace(*src))
	    *src++ = *dup++;

	*src++ = '\'';
	while (*dup) {
	    if (*dup == '\'') {
		*src++ = '\'';
		*src++ = '"';
		*src++ = '\'';
		*src++ = '"';
		*src++ = '\'';
		dup++;
	    } else {
		*src++ = *dup++;
	    }
	}

	*src++ = '\'';
	*src = '\0';
	free(dup_orig);
    }

    return s;
}
#else
#define	quote_length(s) strlen(s)
#define quote_string(s)	(s)
#endif /*QUOTE_POPEN_ARGS */

#define MAX_OPTIONS 40

#define ALL_OPTIONS "\
B:\
C:\
D:\
E:\
F:\
I:\
M:\
O:\
P:\
S\
T\
U:\
V\
X:\
a\
b\
c\
d\
e\
f:\
i\
l\
m\
n\
o:\
p\
q\
s\
t\
v\
x\
"
/* Process the command line options.
 */
static void
process_options(int *pargc, char ***pargv)
{
    int argc, eargc, nargc;
    char **argv, *eargv[MAX_OPTIONS], **nargv;
    int i, c;
    char *s;
#if defined(CPP)
    size_t n;
#endif
#if defined(CPP) && !defined(vms)
    char *tmp;
#endif

    argc = *pargc;
    argv = *pargv;
#ifndef vms			/* this conflicts with use of foreign commands... */
    if ((s = getenv("CPROTO")) != NULL) {
	parse_options(s, MAX_OPTIONS, &eargc, eargv);
	nargv = (char **) xmalloc(((unsigned) (eargc + argc + 1)) * sizeof(char *));
	nargv[0] = argv[0];
	nargc = 1;
	for (i = 0; i < eargc; ++i)
	    nargv[nargc++] = eargv[i];
	for (i = 1; i < argc; ++i)
	    nargv[nargc++] = argv[i];
	nargv[nargc] = NULL;
	argc = nargc;
	argv = nargv;
    }
#endif

#ifdef CPP
    /* Allocate buffer for C preprocessor command line. */
    n = strlen(cpp) + 1;
    for (i = 0; i < argc; ++i) {
	n += quote_length(argv[i]) + 1;		/* add more for possible quoting */
    }
#ifdef	vms
    *(cpp_include = xmalloc(n + argc)) = '\0';
    *(cpp_defines = xmalloc(n + argc)) = '\0';
    *(cpp_undefns = xmalloc(n + argc)) = '\0';
    n += 30;			/* for keywords */
#endif
    *(cpp_opt = (char *) xmalloc(n)) = '\0';
    n += (2 + strlen(CPP) + BUFSIZ);
    *(cpp_cmd = (char *) xmalloc(n)) = '\0';
    cpp_len = n;
#endif

    while ((c = getopt(argc, argv, ALL_OPTIONS)) != EOF) {
	switch (c) {
	case 'I':
#ifdef	vms
	    add2list(cpp_include, optarg);
	    break;
#else /* unix */
	    add_inc_dir(optarg);
#endif
	    /*FALLTHRU */
	case 'D':
#ifdef	vms
	    add2list(cpp_defines, optarg);
	    break;
#endif
	    /*FALLTHRU */
	case 'U':
#ifdef	vms
	    add2list(cpp_undefns, optarg);
	    break;
#else /* UNIX, etc. */
#ifdef CPP
	    tmp = (char *) xmalloc(quote_length(optarg) + 10);
	    sprintf(tmp, " -%c%s", c, optarg);
	    strcat(cpp_opt, quote_string(tmp));
	    free(tmp);
#endif
#endif
	    break;
	case 'n':
	    use_stdinc_dirs = FALSE;
#ifdef CPP
	    tmp = (char *) xmalloc(quote_length("-nostdinc") + 10);
	    sprintf(tmp, " %s", "-nostdinc");
	    strcat(cpp_opt, quote_string(tmp));
	    free(tmp);
#endif
	    break;
	case 'a':
	    func_style = FUNC_ANSI;
	    break;
	case 'B':
	    func_directive = optarg;
	    break;
	case 'b':
	    func_style = FUNC_BOTH;
	    break;
	case 'c':
	    proto_comments = TRUE;
	    break;
	case 'd':
	    define_macro = FALSE;
	    break;
	case 'E':
#ifdef CPP
	    if (strcmp(optarg, "0") == 0) {
		cpp = NULL;
	    } else {
		cpp = optarg;
	    }
#endif
	    break;
	case 'e':
	    extern_out = TRUE;
	    break;
	case 'C':
	case 'F':
	case 'P':
	    s = escape_string(optarg);
	    i = (c == 'C') ? FMT_FUNC_COMMENT :
		((c == 'F') ? FMT_FUNC : FMT_PROTO);

	    fmt[i].decl_spec_prefix = s;
	    while (*s != '\0' && !isalnum(UCH(*s)))
		++s;
	    if (*s == '\0')
		usage();
	    *s++ = '\0';
	    while (*s != '\0' && isalnum(UCH(*s)))
		++s;
	    if (*s == '\0')
		usage();

	    fmt[i].declarator_prefix = s;
	    while (*s != '\0' && !isalnum(UCH(*s)))
		++s;
	    if (*s == '\0')
		usage();
	    *s++ = '\0';
	    while (*s != '\0' && isalnum(UCH(*s)))
		++s;
	    if (*s == '\0')
		usage();

	    fmt[i].declarator_suffix = s;
	    while (*s != '\0' && *s != '(')
		++s;
	    if (*s == '\0')
		usage();
	    *s++ = '\0';

	    fmt[i].first_param_prefix = s;
	    while (*s != '\0' && !isalnum(UCH(*s)))
		++s;
	    if (*s == '\0')
		usage();
	    *s++ = '\0';
	    while (*s != '\0' && *s != ',')
		++s;
	    if (*s == '\0')
		usage();

	    fmt[i].middle_param_prefix = ++s;
	    while (*s != '\0' && !isalnum(UCH(*s)))
		++s;
	    if (*s == '\0')
		usage();
	    *s++ = '\0';
	    while (*s != '\0' && isalnum(UCH(*s)))
		++s;
	    if (*s == '\0')
		usage();

	    fmt[i].last_param_suffix = s;
	    while (*s != '\0' && *s != ')')
		++s;
	    *s = '\0';

	    break;
	case 'f':
	    proto_style = atoi(optarg);
	    if (proto_style < 0 || proto_style > PROTO_ANSI)
		usage();
	    break;
	case 'm':
	    proto_macro = TRUE;
	    break;
	case 'M':
	    macro_name = optarg;
	    break;
	case 'p':
	    promote_param = FALSE;
	    break;
	case 'q':
	    quiet = TRUE;
	    break;
	case 'S':
	    scope_out = SCOPE_STATIC;
	    break;
	case 's':
	    scope_out = SCOPE_ALL;
	    break;
	case 'i':
	    inline_out = TRUE;
	    break;
	case 't':
	    func_style = FUNC_TRADITIONAL;
	    break;
	case 'V':
	    fprintf(stderr, "%s\n", VERSION);
	    exit(EXIT_FAILURE);
	    break;
	case 'v':
	    variables_out = TRUE;
	    break;
	case 'o':
	    if (freopen(optarg, "w", stdout) == 0) {
		perror(optarg);
		exit(EXIT_FAILURE);
	    }
	    break;
	case 'O':
	    if (freopen(optarg, "w", stderr) == 0) {
		perror(optarg);
		exit(EXIT_FAILURE);
	    }
	    break;
#if OPT_LINTLIBRARY
	case 'T':		/* emit typedefs */
	    types_out = TRUE;
	    break;
	case 'l':
	    proto_style = PROTO_LINTLIBRARY;
	    extern_out = FALSE;
	    types_out = TRUE;
	    variables_out = TRUE;
# if !defined(vms) && !defined(MSDOS)
	    (void) strcat(cpp_opt, " -C");	/* pass-through comments */
# endif
	    break;
	case 'X':
	    extern_in = (unsigned) atoi(optarg);
	    do_tracking = TRUE;
	    if ((int) extern_in < 0)
		usage();
	    break;
#endif /* OPT_LINTLIBRARY */
	case 'x':
	    extern_in = MAX_INC_DEPTH;
	    break;
	default:
	    usage();
	}
    }

#ifdef	vms
    add_option("includes", cpp_include);
    add_option("define", cpp_defines);
    add_option("undefine", cpp_undefns);
#endif

    *pargc = argc;
    *pargv = argv;
}

static void add_stdinc_dir(void)
{
#if defined(MSDOS) && defined(__TURBOC__)
    add_inc_dir("/tc/include");
#elif defined(vms)
    add_inc_dir("sys$library:");
#else
    add_inc_dir("/usr/include");
#endif
    return;
}

int
main(int argc, char *argv[])
{
    int i;
    FILE *inf;
#ifdef NO_LEAKS
    char *argv0;
#endif

#ifdef __EMX__
    /* Expand file wild cards. */
    _wildcard(&argc, &argv);
#endif

    add_inc_dir(CURRENT_DIR);

    /* Get the program name from the 0th argument, stripping the pathname
     * for readability.
     */
    progname = xstrdup(argv[0]);
#ifdef NO_LEAKS
    argv0 = progname = argv0;
#endif
#ifdef vms
    for (i = strlen(progname) - 1; i >= 0; i--) {
	if (progname[i] == SQUARE_R
	    || progname[i] == ':') {
	    progname += (i + 1);
	    break;
	} else if (progname[i] == '.') {
	    progname[i] = '\0';
	}
    }
#else
    for (i = (int) strlen(progname) - 1; i >= 0; i--) {
	if (is_path_sep(progname[i])) {
	    progname += (i + 1);
	    break;
# if defined(MSDOS) || defined(OS2)
	} else if (progname[i] == '.') {
	    progname[i] = '\0';
# endif
	}
    }
#endif
    argv[0] = progname;		/* do this so getopt is consistent with us */

    process_options(&argc, &argv);

    if (use_stdinc_dirs) {
      add_stdinc_dir();
    }

#if OPT_LINTLIBRARY
    if (lintLibrary()) {
	put_string(stdout, "/* LINTLIBRARY */\n");
	switch (func_style) {
	case FUNC_ANSI:
	case FUNC_BOTH:
	    lint_shadowed = TRUE;	/* e.g., ctype.h */
	    proto_style = PROTO_ANSI_LLIB;
	    break;
	}
	func_style = FUNC_NONE;
    } else if (func_style == FUNC_UNKNOWN)
	func_style = FUNC_NONE;
#endif

    if (proto_macro && define_macro) {
	printf("#if __STDC__ || defined(__cplusplus)\n");
	printf("#define %s(s) s\n", macro_name);
	printf("#else\n");
	printf("#define %s(s) ()\n", macro_name);
	printf("#endif\n\n");
    }

    init_parser();
    if (optind == argc) {
	if (func_style != FUNC_NONE) {
	    proto_style = PROTO_NONE;
	    variables_out = FALSE;
	    file_comments = FALSE;
	}
	process_file(stdin, "stdin");
	pop_file(FALSE);
    } else {
	if (!optind)
	    optind++;
	for (i = optind; i < argc; ++i) {
#ifdef CPP
# if defined(CPP_DOES_ONLY_C_FILES)
	    /*
	     * GCC (and possibly other compilers) don't pass-through the ".l"
	     * and ".y" files to the C preprocessor stage.  Fix this by
	     * temporarily linking the input file to a temporary-file with a
	     * ".c" suffix.
	     *
	     * Of course, this solution assumes that the input directory is
	     * writeable.
	     */
	    char temp[BUFSIZ];
	    char *s = strcpy(temp, argv[i]);
#  if HAVE_LINK
	    int len = (int) strlen(temp);
	    s += len - 1;
	    if ((len > 2)
		&& (s[-1] == '.')
		&& (*s == 'l' || *s == 'y')) {
		while (s != temp && s[-1] != '/')
		    s--;
		(void) strcpy(s, "XXXXXX.c");
		call_mktemp(temp);
		if (link(argv[i], temp) < 0)
		    (void) strcpy(temp, argv[i]);
	    }
#  endif
#  define FileName temp
# else
#  define FileName argv[i]
#  ifdef vms
	    char temp[BUFSIZ];
	    (void) strcpy(temp, FileName);
#  endif
# endif
	    if (func_style == FUNC_NONE && cpp != NULL) {
#ifdef vms
		/*
		 * Assume the 'cpp' command contains a "%s" for the name of
		 * the file that we're writing to.
		 */
		sprintf(cpp_cmd, cpp,
			call_mktemp(strcpy(temp, "sys$scratch:XXXXXX.i")));
		sprintf(cpp_cmd + strlen(cpp_cmd), "%s %s", cpp_opt, FileName);
		system(cpp_cmd);
		inf = fopen(temp, "r");
#else
		if (cpp_len < (strlen(cpp) + strlen(cpp_opt) +
			       strlen(FileName) + 100)) {
		    cpp_len = (strlen(cpp) + strlen(cpp_opt) +
			       strlen(FileName) + 100);
		    cpp_cmd = realloc(cpp_cmd, cpp_len);
		}
		sprintf(cpp_cmd, "%s%s %s", cpp, cpp_opt, FileName);
		if (quiet)
		    strcat(cpp_cmd, " 2>/dev/null");

		inf = popen(cpp_cmd, "r");
#endif
		if (inf == NULL || ferror(inf) || feof(inf)) {
		    fprintf(stderr, "%s: error running %s\n", progname,
			    cpp_cmd);
		    continue;
		}
	    } else {
		if ((inf = fopen(argv[i], "r")) == NULL) {
		    fprintf(stderr, "%s: cannot read file %s\n", progname,
			    argv[i]);
		    continue;
		}
	    }
#else
	    if ((inf = fopen(argv[i], "r")) == NULL) {
		fprintf(stderr, "%s: cannot read file %s\n", progname, argv[i]);
		continue;
	    }
#endif
	    process_file(inf, argv[i]);
#ifdef CPP
	    if (func_style == FUNC_NONE && cpp != NULL) {
#ifdef vms
		fclose(inf);
#else
		pclose(inf);
#endif
#if defined(CPP_DOES_ONLY_C_FILES) || defined(vms)
		if (strcmp(argv[i], temp)) {
		    (void) unlink(temp);
		}
		pop_file(TRUE);
#endif
	    } else {
		pop_file(FALSE);
	    }
#else /* no CPP defined */
	    pop_file(FALSE);
#endif
	}
    }

    if (proto_macro && define_macro) {
	printf("\n#undef %s\n", macro_name);
    }
#ifdef NO_LEAKS
# ifdef CPP
    if (cpp_opt != 0)
	free(cpp_opt);
    if (cpp_cmd != 0)
	free(cpp_cmd);
#  ifdef vms
    if (cpp_include != 0)
	free(cpp_include);
    if (cpp_defines != 0)
	free(cpp_defines);
    if (cpp_undefns != 0)
	free(cpp_undefns);
#  endif
# endif
    if (inc_dir != 0) {
	while (num_inc_dir-- > 0) {
	    free(inc_dir[num_inc_dir]);
	}
	free(inc_dir);
	inc_dir = 0;
    }
    free_lexer();
    free_parser();
# if OPT_LINTLIBRARY
    free_lintlibs();
# endif

# ifdef DOALLOC
    show_alloc();
# endif
    free(argv0);
#endif

    return EXIT_SUCCESS;
}

/* Intercept 'exit()' for debugging.  (The Linux libc uses malloc/free in
 * 'exit()', so we cannot get a trace unless we resort to this hack ;-)
 */
#if HAVE_LIBDBMALLOC
#undef exit
void
ExitProgram(int code)
{
    extern int malloc_errfd;	/* FIXME: should be in dbmalloc.h */
    malloc_dump(malloc_errfd);
    exit(code);
}
#endif
